### PR TITLE
util: inline get_limit_at_size

### PR DESCRIPTION
**Pull request by odeits**
removes need to use Iterator::count and allows to not call Vec::truncate
if NO_LIMIT

Fixes pingcap/tikv#2400

### CODE REVIEW

File: src/util/mod.rs [hunk 1]
```diff
@@ -50,35 +51,20 @@ mod thread_metrics;
...
-    let mut limit = 1;
-    for e in iter {
-        size += Message::compute_size(e) as u64;
-        if size > max {
-            break;
+pub fn limit_size<T: Message + Clone>(entries: &mut Vec<T>, max: u64) {
+    if max != NO_LIMIT {
+        let mut limit = 0;
+        let mut size = 0;
+        for e in entries.into_iter() {
```
Comment (BusyJay) [inline src/util/mod.rs]: You can use `entries.iter().take_while().count()`.

File: src/util/mod.rs [hunk 2]
```diff
@@ -50,35 +51,20 @@ mod thread_metrics;
...
+            size += Message::compute_size(e) as u64;
+            if size > max {
+                break;
+            }
+            limit += 1;
         }
-        limit += 1;
+        limit = cmp::max(limit, 1);
```
Comment (BusyJay) [inline src/util/mod.rs]: Maybe check if `entries.len() <= 1` and return early is better. This method can be called frequently with only one entry in vector.

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay

**Comment by BusyJay** [comment]
LGTM, please format your code.

**Comment by siddontang** [comment]
Any update @odeits?

**Comment by odeits** [comment]
I am having trouble compiling. I used the script https://github.com/pingcap/docs/blob/master/scripts/build_rocksdb.sh 
to install rocksdb


```
error: failed to run custom build command for `librocksdb_sys v0.1.0 (https://github.com/pingcap/rust-rocksdb.git#229732cb)`
process didn't exit successfully: `/tikv/target/debug/build/librocksdb_sys-3713e0023fbb987f/build-script-build` (exit code: 101)
--- stdout
cargo:rustc-link-lib=rocksdb
TARGET = Some("x86_64-unknown-linux-gnu")
OPT_LEVEL = Some("0")
TARGET = Some("x86_64-unknown-linux-gnu")
HOST = Some("x86_64-unknown-linux-gnu")
TARGET = Some("x86_64-unknown-linux-gnu")
TARGET = Some("x86_64-unknown-linux-gnu")
HOST = Some("x86_64-unknown-linux-gnu")
CXX_x86_64-unknown-linux-gnu = None
CXX_x86_64_unknown_linux_gnu = None
HOST_CXX = None
CXX = None
HOST = Some("x86_64-unknown-linux-gnu")
TARGET = Some("x86_64-unknown-linux-gnu")
HOST = Some("x86_64-unknown-linux-gnu")
CXXFLAGS_x86_64-unknown-linux-gnu = None
CXXFLAGS_x86_64_unknown_linux_gnu = None
HOST_CXXFLAGS = None
CXXFLAGS = None
DEBUG = Some("true")
running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-m64" "-std=c++11" "-Wall" "-Wextra" "-o" "/tikv/target/debug/build/librocksdb_sys-a40eeeb2edcbb581/out/crocksdb/c.o" "-c" "crocksdb/c.cc"
cargo:warning=crocksdb/c.cc:32:37: fatal error: rocksdb/utilities/debug.h: No such file or directory
cargo:warning= #include "rocksdb/utilities/debug.h"
cargo:warning=                                     ^
cargo:warning=compilation terminated.
exit code: 1

--- stderr
thread 'main' panicked at '

Internal error occurred: Command "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-m64" "-std=c++11" "-Wall" "-Wextra" "-o" "/tikv/target/debug/build/librocksdb_sys-a40eeeb2edcbb581/out/crocksdb/c.o" "-c" "crocksdb/c.cc" with args "c++"
 did not execute successfully (status code exit code: 1).

', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.0/src/lib.rs:1645
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:371
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:511
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:495
   7: cc::fail
             at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.0/src/lib.rs:1645
   8: cc::Build::compile
             at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/cc-1.0.0/src/lib.rs:760
   9: build_script_build::main
             at ./build.rs:23
  10: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
  11: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:433
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  12: main
  13: __libc_start_main
  14: <unknown>

Build failed, waiting for other jobs to finish...
error: build failed
make: *** [test] Error 101
```

**Comment by BusyJay** [comment]
I think just `make format` should fix CI. Please make sure you are using rustfmt 0.2.1-nightly. Or you can adjust your code manually according to the CI output.

**Comment by siddontang** [comment]
@odeits 

We don't use RocksDB 5.6.1 now, the building script is out of date, I will remove it later.

**Comment by siddontang** [comment]
Hi @odeits 

CI still failed, you can install rustfmt 0.2.1-nightly and run `make format`.

**Comment by odeits** [comment]
@siddontang Sorry I have been having trouble getting `make format` to work. Issues should be resolved now.

**Pull request merged**